site/content/extend/plugins/signing-plugins.md

```diff
@@ -0,0 +1,151 @@
+
+## Key Generation
+Public and private key pairs are needed to sign and verify plugins. Private key is used for signing and should be kept in the Vault. Public key is used for verification and can be distributed freely. To generate a key pair, run the following command:
+
+
+ ```console
+ > gpg --full-generate-key
+
+ Please select what kind of key you want:
+    (1) RSA and RSA (default)
+    (2) DSA and Elgamal
+    (3) DSA (sign only)
+    (4) RSA (sign only)
+ Your selection? 1
+
+ RSA keys may be between 1024 and 4096 bits long.
+ What keysize do you want? (2048) 3072
+
+ Requested keysize is 3072 bits
+
+ Please specify how long the key should be valid.
+          0 = key does not expire
+       <n>  = key expires in n days
+       <n>w = key expires in n weeks
+       <n>m = key expires in n months
+       <n>y = key expires in n years
+ Key is valid for? (0) 0
+
+ Key expires at ...
+
+ Is this correct? (y/N) y
+
+ GnuPG needs to construct a user ID to identify your key.
+ Real name: Mattermost Inc
+
+ Email address: info@mattermost.com
+
+ Comment:
+
+ You selected this USER-ID:
+     "Mattermost Inc <info@mattermost.com>"
+ Change (N)ame, (C)omment, (E)mail or (O)kay/(Q)uit? O
+
+ ```
+ Note: key size should be at least 3072 bits.
+
+### Exporting the Private Key
+Find the ID of your private key first. The ID is a hexadecimal number.
+```console
+> gpg --list-secret-keys
+```
+This is your private key and you should keep it secret. Replace `XXXXXXXX` with your hexadecimal key ID.
+```console
+> gpg --export-secret-keys XXXXXXXX > ./my-priv-key
+```
+
+
+### Exporting the Public Key
+Find the ID of your public key first. The ID is a hexadecimal number.
+```console
+> gpg --list-keys
+```
+Replace `XXXXXXXX` with your hexadecimal key ID.
+```console
+> gpg --export XXXXXXXX > ./my-pub-key
+```
+
+
+### Importing the Key
+If you already have a public and private key pair, you can import them to the GPG.
+```console
+> gpg --import ./my-priv-gpg-key
+> gpg --import ./my-pub-gpg-key
+```
+
+
+## Plugin Signing
+For plugin signing you have to know the hexadecimal ID of the private key. Let's assume you want to sign `com.mattermost.demo-plugin-0.1.0.tar.gz` file. Replace `XXXXXXXX` with your hexadecimal key ID and run:
+```console
+> gpg -u XXXXXXXX --verbose --personal-digest-preferences SHA256 --detach-sign com.mattermost.demo-plugin-0.1.0.tar.gz
+```
+This command will generate `com.mattermost.demo-plugin-0.1.0.tar.gz.sig`, which is the signature of your plugin.
+
+## Plugin Verification
+Mattermost server will verify plugin signatures downloaded from Mattermost Marketplace. Mattermost public key is shipped with the server. If needed one can download and verify plugins from custom marketplace as well. To do this, custom public key has to be added on the server. For this purpose, start the Mattermost server and run the following command:
```

Comment:
Let's use the term `plugin marketplace` to avoid talking about the Mattermost-hosted one, specifically. Also, I proposed a change above that gives the context up front, so we can trim this to just focus on the "how".